File: leetcode/No63.unique-paths-II.js
/**
 * Difficulty:
 * Medium
 *
 * Desc:
 * Follow up for "Unique Paths":
 * Now consider if some obstacles are added to the grids. How many unique paths would there be?
 * An obstacle and empty space is marked as 1 and 0 respectively in the grid.
 *
 * Example:
 * There is one obstacle in the middle of a 3x3 grid as illustrated below.
 * [
 *    [0,0,0],
 *    [0,1,0],
 *    [0,0,0]
 * ]
 * The total number of unique paths is 2.
 *
 * 依旧是走棋盘，但棋盘上有些位置会有阻碍（用 1 表示），使得道路不通
 */

/**
 * 和上一题的思路一致，使用动态规划
 * 但要注意特殊位置的问题
 * 当走到障碍处之后，之前积累的路径数情况，且障碍点对其之后的点贡献的路径数为 0
 */

/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
  var steps = [];
  for (var i = 0; i < obstacleGrid.length; i += 1) {
    var col = obstacleGrid[i];
    for (var j = 0; j < col.length; j += 1) {
      var key = `${i}${j}`;
      if (col[j] === 1) {
        steps[key] = 0;
        continue;
      }
      if (i === 0 && j === 0) {
        steps[key] = 1;
        continue;
      }
      if (i === 0 || j === 0) {
        var preKey = `${i === 0 ? i : i - 1}${j === 0 ? j : j - 1}`;
        steps[key] = steps[preKey];
        continue;
      }
      steps[key] = steps[`${i}${j - 1}`] + steps[`${i - 1}${j}`];
    }
  }
  return steps[`${obstacleGrid.length - 1}${obstacleGrid[0].length - 1}`];
};

console.log(uniquePathsWithObstacles([ [ 0 ] ]));
console.log(uniquePathsWithObstacles(
  [
    [0,0,0],
    [0,1,0],
    [0,0,0]
  ]
));

File: leetcode/C#/No63.unique-paths-II.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * A robot is located at the top-left corner of a m x n grid (marked 'Start' in the diagram below).
 * The robot can only move either down or right at any point in time.
 * The robot is trying to reach the bottom-right corner of the grid (marked 'Finish' in the diagram below).
 * Now consider if some obstacles are added to the grids. How many unique paths would there be?
 * An obstacle and empty space is marked as 1 and 0 respectively in the grid.
 * Note: m and n will be at most 100.
 *
 * Example 1:
 * Input:
 * [
 *  [0,0,0],
 *  [0,1,0],
 *  [0,0,0]
 * ]
 * Output: 2
 * Explanation:
 * There is one obstacle in the middle of the 3x3 grid above.
 * There are two ways to reach the bottom-right corner:
 * 1. Right -> Right -> Down -> Down
 * 2. Down -> Down -> Right -> Right
*/

public class Solution {
    private int[][] dirs = new int[2][] {
        new int[2] { 0, 1 },
        new int[2] { 1, 0 }
    };

    private int Dfs(int i, int j, int[][] obstacleGrid, int[,] cache) {
        if (obstacleGrid[i][j] == 1) {
            cache[i, j] = -1;
            return 0;
        }
        if (i == obstacleGrid.Length - 1 && j == obstacleGrid[0].Length - 1) {
            return 1;
        }
        if (cache[i, j] > 0) {
            return cache[i, j];
        }

        int res = 0;
        foreach (int[] dir in dirs) {
            int row = i + dir[0];
            int col = j + dir[1];
            if (row < obstacleGrid.Length && col < obstacleGrid[0].Length && obstacleGrid[row][col] == 0) {
                res += Dfs(row, col, obstacleGrid, cache);
            }
        }

        cache[i, j] = res == 0 ? -1 : res;
        return res;
    }

    public int UniquePathsWithObstacles(int[][] obstacleGrid) {
        int[,] cache = new int[obstacleGrid.Length, obstacleGrid[0].Length];
        return Dfs(0, 0, obstacleGrid, cache);
    }
}